### PR TITLE
MHA: Stricter input validation

### PR DESCRIPTION
**Pull request by danthe3rd**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #594
* #593
* __->__ #592

### CODE REVIEW

File: xformers/ops/fmha/common.py [hunk 1]
```diff
@@ -99,6 +99,43 @@ def normalize_bmhk(self) -> Tuple[int, ...]:
             self.value = self.value.unsqueeze(2)
         return output_shape
 
+    def validate_bmhk(self) -> None:
+        qkv = (self.query, self.key, self.value)
+        if tuple(x.ndim for x in qkv) != (4, 4, 4):
+            raise ValueError(
+                f"Query/Key/Value should have BMHK format.\n"
```
Comment (fmassa) [inline xformers/ops/fmha/common.py:106]: This error could be misleading. The user can pass a `BMK` tensor as well, and will be confused as to why this mentions that only 4d inputs are valid

File: xformers/ops/fmha/__init__.py [hunk 2]
```diff
@@ -321,6 +322,7 @@ def _memory_efficient_attention_forward_requires_grad(
         raise ValueError(
             f"xformers.memory_efficient_attention: Operator {op.NAME} does not support this input"
         )
+    inp.validate_bmhk()
```
Comment (fmassa) [inline xformers/ops/fmha/__init__.py:325]: I would have seen those checks right in the beginning of the function, just before `normalize_bmhk`, this way we could also capture the fact that the tensors can be 3d or 4d

File: xformers/ops/fmha/common.py [hunk 3]
```diff
@@ -99,6 +99,43 @@ def normalize_bmhk(self) -> Tuple[int, ...]:
             self.value = self.value.unsqueeze(2)
         return output_shape
 
+    def validate_bmhk(self) -> None:
```
Comment (fmassa) [inline xformers/ops/fmha/common.py:102]: if you move this to the beginning of the function, I would maybe call it `validate_inputs`?
